core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -152,7 +152,9 @@ class GroupMetadataManager(brokerId: Int,
 
   def isPartitionLoading(partition: Int) = inLock(partitionLock) { loadingPartitions.contains(partition) }
 
-  def partitionFor(groupId: String): Int = Utils.abs(groupId.hashCode) % groupMetadataTopicPartitionCount
+  def maybeUpdateGroupMetadataTopicPartitionCount() = _groupMetadataTopicPartitionCount = getGroupMetadataTopicPartitionCount
```

Comment:
Inside `getGroupMetadataTopicPartitionCount`, we use the default number of partitions if the topic hasn't been created yet. Would it make more sense to let `groupMetadataTopicPartitionCount` be an `Option[Int]` and wait until topic creation before trying to use it?

Also, it may be helpful to log an info message when the number of partitions changes.

Reply:
Is it okay to still use `Int`  as the type of `groupMetadataTopicPartitionCount` but eagerly assign it  as `getGroupMetadataTopicPartitionCount.orElse(Some(config.offsetsTopicNumPartitions))` during the instance construction? We only need to ensure the partition count could be updated after a manual change via issuing CreateTopics or AddPartitions request by AdminClient.

We could only reflect the count in UpdateMetadata request which could lead to invalid values returned from `partitionFor` if it is called before processing  UpdateMetadata. Does it make sense?

Reply:
Basically I think it is dangerous to rely on the partition count until the topic is created. So we could raise an exception if we try to use the default value before topic creation. Basically something like this:
```scala
zkClient.getTopicPartitionCount(Topic.GROUP_METADATA_TOPIC_NAME)
  .getOrElse(throw CoordinatorNotAvailableException())
```
The most important thing is to get the update hooks into the right place so that we do not get the count wrong. Now that I'm thinking about it, we probably need the following:

1. Inside `GroupMetadataManager.scheduleLoadGroupAndOffsets`, we should verify that the partition is valid. If the partition is larger than the count we have locally, we should reload the partition count.
2. As discussed above, when we receive a metadata update for a partition that is larger than the count we have locally, we should reload.

Finally, it looks like we need to fix the ordering here of creation inside `KafkaApis`:
```scala
          val partition = groupCoordinator.partitionFor(findCoordinatorRequest.coordinatorKey)
          val metadata = getOrCreateInternalTopic(GROUP_METADATA_TOPIC_NAME, request.context.listenerName)
          (partition, metadata)
```
This is racy in the presence of manual topic creation. It would make more sense to call `getOrCreateInternalTopic` before calling `partitionFor`. 

I am not 100% sure it is necessary to be super strict about this, but I think it's worth starting strict and only backing off if we find some major problems.